core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,379 @@
+import com.graphhopper.routing.ev.EncodedValue;
+import com.graphhopper.routing.ev.EncodedValueLookup;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.util.EdgeIteratorState;
+import com.graphhopper.util.GHUtility;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.Polygon;
+import org.codehaus.commons.compiler.CompileException;
+import org.codehaus.commons.compiler.Location;
+import org.codehaus.commons.compiler.io.Readers;
+import org.codehaus.janino.Scanner;
+import org.codehaus.janino.*;
+import org.codehaus.janino.util.DeepCopier;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicLong;
+
+class ExpressionBuilder {
+    private static final AtomicLong longVal = new AtomicLong(1);
+    static final String IN_AREA_PREFIX = "in_area_";
+    private static final Set<String> allowedNames = new HashSet<>(Arrays.asList("edge", "Math"));
+    private static final boolean JANINO_DEBUG = Boolean.getBoolean(Scanner.SYSTEM_PROPERTY_SOURCE_DEBUGGING_ENABLE);
+
+    // TODO without a cache we get X% slower routing and Y% slower routingLM8. CH requests and preparation is unaffected
+    // as cached weighting from preparation is used. This cache ensures that the first Weighting classes, typically the
+    // ones specified in the profiles, are never removed regardless of how frequent other Weightings are created and accessed.
+    // We only need to synchronize the get and put methods alone. E.g. we do not care for the race condition where
+    // two identical classes are requested and one of them is overwritten. We could use CachingJavaSourceClassLoader from
+    // Janino but 1. we would need to use a single compiler across threads and 2. the statements (for priority and speed)
+    // are still unnecessarily created.
+    private static final int CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.cache_size", 10);
+    // TODO perf compare with ConcurrentHashMap, but I guess, if there is a difference at all, it is not big for small maps
+    private static final Map<String, Class> CACHE = Collections.synchronizedMap(new HashMap<>(CACHE_SIZE));
+
+    // Introduce a dynamic cache to remember different Weighting classes, but throw away less frequently used classes.
+    // Use accessOrder==true to remove oldest accessed entry, not oldest inserted.
+    private static final int DYN_CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.dynamic_cache_size", 1000);
+    private static final Map<String, Class> DYN_CACHE = Collections.synchronizedMap(
+            new LinkedHashMap<String, Class>(DYN_CACHE_SIZE, 0.75f, true) {
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    return size() > DYN_CACHE_SIZE;
+                }
+            });
+
```

Comment:
I would put a private constructor here to make it clear that this class is just a collection of static methods (if this is what it is supposed to be).